src/Features/Core/Portable/CodeRefactorings/MoveType/AbstractMoveTypeService.Editor.cs

```diff
@@ -0,0 +1,272 @@
+
+                var membersToRemove = GetMembersToRemove(root, typeNode, removeDescendentsOfSelf: _makeTypePartial);
+                foreach (var member in membersToRemove)
+                {
+                    documentEditor.RemoveNode(member, SyntaxRemoveOptions.KeepNoTrivia);
+                }
+
+                var memberToAdd = GetMemberToAdd(documentEditor.Generator, newDocumentName);
+                if (memberToAdd != null)
+                {
+                    documentEditor.AddMember(typeNode, memberToAdd);
+                }
+
+                var modifiedRoot = documentEditor.GetChangedRoot();
+
+                // add an empty document to solution, so that we'll have options from the right context.
+                var solutionWithNewDocument = projectToBeUpdated.Solution.AddDocument(newDocumentId, newDocumentName, string.Empty/*, folders, fullFilePath*/);
+
+                // update the text for the new document
+                solutionWithNewDocument = solutionWithNewDocument.WithDocumentSyntaxRoot(newDocumentId, modifiedRoot, PreservationMode.PreserveIdentity);
+
+                // get the updated document, perform clean up like remove unused usings.
+                var newDocument = solutionWithNewDocument.GetDocument(newDocumentId);
+                return await CleanUpDocumentAsync(newDocument, cancellationToken).ConfigureAwait(false);
+            }
+
+            private async Task<Solution> UpdateSourceDocumentAsync(
+                Document sourceDocument, TTypeDeclarationSyntax typeNode, CancellationToken cancellationToken)
+            {
+                var documentEditor = await DocumentEditor.CreateAsync(sourceDocument, cancellationToken).ConfigureAwait(false);
+
+                AddPartialModifiersToTypeChain(documentEditor, typeNode);
+
+                if (!_makeTypePartial)
+                {
+                    documentEditor.RemoveNode(typeNode, SyntaxRemoveOptions.KeepNoTrivia);
+                }
+
+                var updatedDocument = documentEditor.GetChangedDocument();
+                // TODO: make this optional -- shouldn't touch other parts of source document unless asked to. 
+                return await CleanUpDocumentAsync(updatedDocument, cancellationToken).ConfigureAwait(false);
+            }
+
+            private SyntaxNode GetMemberToAdd(SyntaxGenerator generator, string newDocumentName)
+            {
+                if (!_makeOuterTypesPartial)
+                {
+                    return null;
+                }
+
+                // we recognize filenames of that follow the pattern "OuterClass.InnerClass.cs" 
+                // and generate code for this file, which will look like so:
+                // a partial definition of OuterClass and a nested private InnerClass.
+                var nameParts = Path.GetFileNameWithoutExtension(newDocumentName).Split(new char[] { '.' });
+                if (nameParts == null || nameParts.Count() != 2)
```

Comment:
1. and where do we stop? Notice that since we would get the name from the user via the dialog, its easy to construct pathological cases that are deeply nested. I do not see any value in supporting them. I'm not saying 2 is the right number either, but still thinking on what's the best solution here..

2. Another cool thing to do here:  In A.B.C.cs, `A.B` could be the containing namespace and `C` the typename? 

Talking with @Pilchie and @kuhlenh , it seems the most straight forward scenario is the most wanted feature set here. So, I'm thinking about adding these incrementally.